File: backend/src/services/articleService.ts
import ArticleBuilder from "../builders/articleBuilder.js";
import { db, articles } from "../db/client.js";
import { type Article } from "../db/schema.js";
import { eq } from "drizzle-orm";
import { searchService } from "./searchService.js";

export default class ArticleService {
  async list(categoryId?: number): Promise<Article[]> {
    if (categoryId) {
      return db.select().from(articles).where(eq(articles.categoryId, categoryId));
    }
    return db.select().from(articles);
  }

  async getById(id: number): Promise<Article | undefined> {
    const [result] = await db.select().from(articles).where(eq(articles.id, id));
    return result;
  }

  async create(input: {
    name: string;
    description?: string;
    imageUrl: string;
    categoryId: number;
    stock: number;
    needsRepair?: boolean;
  }): Promise<Article> {
    const newArticle = new ArticleBuilder()
      .setName(input.name)
      .setDescription(input.description ?? "")
      .setImageUrl(input.imageUrl)
      .setCategoryId(input.categoryId)
      .setStock(input.stock)
      .setNeedsRepair(input.needsRepair ?? false)
      .build();

    const [created] = await db.insert(articles).values(newArticle).returning();
    searchService.invalidate();
    return created;
  }

  async update(
    id: number,
    input: { name: string; description?: string; imageUrl: string; categoryId: number; stock: number; needsRepair?: boolean },
  ): Promise<Article | undefined> {
    const updatedArticle = new ArticleBuilder()
      .setName(input.name)
      .setDescription(input.description ?? "")
      .setImageUrl(input.imageUrl)
      .setCategoryId(input.categoryId)
      .setStock(input.stock)
      .setNeedsRepair(input.needsRepair ?? false)
      .build();

    const [updated] = await db
      .update(articles)
      .set(updatedArticle)
      .where(eq(articles.id, id))
      .returning();
    searchService.invalidate();
    return updated;
  }
}

File: agente/src/application/services/article_service.py
from ...domain.entities import Article, NewArticle
from ...domain.ports import IArticleRepo


class ArticleService:
    def __init__(self, repo: IArticleRepo) -> None:
        self._repo = repo

    def list_all(self, category_id: int | None = None) -> list[Article]:
        return self._repo.list_all(category_id)

    def get_by_id(self, id: int) -> Article:
        return self._repo.get_by_id(id)

    def create(
        self,
        name: str,
        category_id: int,
        stock: int = 0,
        description: str = "",
        image_url: str = "",
        needs_repair: bool = False,
    ) -> Article:
        return self._repo.create(NewArticle(
            name=name,
            categoryId=category_id,
            stock=stock,
            description=description,
            imageUrl=image_url,
            needsRepair=needs_repair,
        ))

    def update(
        self,
        id: int,
        name: str,
        category_id: int,
        stock: int = 0,
        description: str = "",
        image_url: str = "",
        needs_repair: bool = False,
    ) -> Article:
        return self._repo.update(id, NewArticle(
            name=name,
            categoryId=category_id,
            stock=stock,
            description=description,
            imageUrl=image_url,
            needsRepair=needs_repair,
        ))

    def delete(self, id: int) -> None:
        self._repo.delete(id)
